### PR TITLE
bz 1427254  Kibana Index Mode

### PR DESCRIPTION
**Pull request by jcantrill**
This is a result of a discussion I had with @portante and ops regarding https://bugzilla.redhat.com/show_bug.cgi?id=1427254

Justification is ops users already have access to the entire cluster and this would allow them to share dashboards and visualizations unrestricted.

### CODE REVIEW

**Comment by richm** [comment]
This means all ops users are the same user.  This means if an admin makes customizations, someone else can overwrite those customizations?  All ops users are _not_ the same actual user.  There may be customers who will be perturbed because some other admin can wipe out their customizations.

**Comment by portante** [comment]
@richm, ops users can still create their own dashboards with a different name.  We have hundreds of dashboards on our Kibana instance.  Is it not reasonable to assume that what one ops user needs can be shared by all, and they can agree using dashboard names to avoid conflicts?

**Comment by richm** [comment]
@portante ok - but that means admins will need some agreement and convention about which dashboards are shared, and which are private, rather than being strictly separated by username/userid.  And that will not be consistent with the way normal users have their own config/dashboards.

**Comment by jcantrill** [comment]
@richm Agree all ops users are not the same but most likely all ops users will require visualizations to the same data.  As @portante  suggested, they can come to an agreement on a naming convention.  Aside from figuring out how to 'share' across our design, this is the best scenario to enable sharable dashboards for ops.  The only thing we might add is a config setting that allows you to make it work in a legacy fashion, but generally I see this as a reasonable change.

**Comment by richm** [comment]
Proposed documentation:
```
There is a change from previous versions.
If you login to Kibana as an administrative user, **all of your configs and dashboards are shared with all other administrative users**.  You will need to come up with some sort of naming convention and agreement to share and not overwrite the work of other admin users.
This does not apply if you login to Kibana as a non-admin user.  All of your configs and dashboards will continue to be private to your userid.
```

**Comment by jcantrill** [comment]
One drawback I can think of for this arrangement is users who might want to separate their 'admin' experience from their 'user' experience; they will ONLY be able to use their 'user' profile if some one removes their cluster-admin role.  Do we care? Is this a concern?

**Comment by portante** [comment]
@richm, perhaps explain that this facilitates sharing of dashboards across admin users who are able to see all the data anyways, while preserving the nature of multi-tenancy for non-admin users.  "With great power comes great responsibility."

**Comment by richm** [comment]
There are more elegant ways to share configs/dashboards.  However, if this functionality needed urgently, we can go with this fix and revisit this issue at a later date.

**Comment by jcantrill** [comment]
@richm @portante  I fixed up but will need to rebase once #66 merges. Everything is all jumbled here.

**Comment by jimmidyson** [comment]
retest this please

**Comment by jcantrill** [comment]
[test]

**Comment by jimmidyson** [comment]
retest this please

**Comment by jimmidyson** [comment]
@jcantrill Tests passed now that AWS has sorted stuff out. Will leave it to you to merge.

File: README.md [hunk 1]
```diff
@@ -2,9 +2,16 @@
 This is an OpenShift plugin to ElasticSearch to:
 
 * Dynamically update the SearchGuard ACL based on a user's name
-* Transform kibana index requests to support multitenant deployments
+* Transform kibana index requests to support multitenant deployments for non-operations users.
 * Seed the Searchguard index `config`, `roles`, `rolesmapping`, and `actiongroups` types
 
+*Note:*
+Previous versions of this plugin created a Kibana profile for each user regardless of their role.
+As of v2.4.4.2, any users that has been given the OpenShift cluster-admin role will share the same
```
Comment (portante) [inline README.md]: s/has/have

File: pom.xml [hunk 2]
```diff
@@ -79,7 +79,7 @@
     <!-- used by vendor builds to use vendor version of searchguard plugins -->
     <version.suffix />
   </properties>
-
+
```
Comment (portante) [inline pom.xml]: Seems like an unnecessary change ...

File: src/main/java/io/fabric8/elasticsearch/plugin/acl/DynamicACLFilter.java [hunk 3]
```diff
@@ -78,407 +78,424 @@
...
-		String kspass = settings.get(SG_CLIENT_KS_PASS, DEFAULT_SG_CLIENT_KS_PASS);
-		String tspass = settings.get(SG_CLIENT_TS_PASS, DEFAULT_SG_CLIENT_TS_PASS);
-		String kstype = settings.get(SG_CLIENT_KS_TYPE, DEFAULT_SG_CLIENT_KS_TYPE);
-		String tstype = settings.get(SG_CLIENT_TS_TYPE, DEFAULT_SG_CLIENT_TS_TYPE);
-
-		Settings.Builder settingsBuilder = Settings
-                .builder()
-                .put("path.home", ".")
-                .put("path.conf", ".")
+public class DynamicACLFilter extends RestFilter implements ConfigurationSettings {
```
Comment (portante) [inline src/main/java/io/fabric8/elasticsearch/plugin/acl/DynamicACLFilter.java]: This looks like a huge change just for whitespace mostly.  Why such a large amount of change like that?

File: README.md [hunk 4]
```diff
@@ -2,9 +2,16 @@
...
+Kibana index to store configurations and dashboards.  This change makes it possible to share dashboards
+between operation users.  It is highly recommended that operations teams establish agreements and
+naming conventions so users do not overwrite each others work.
```
Comment (portante) [inline README.md]: So we need to consider this change a bit in terms of how one would migrate to start using it.

What happens if we deploy this where each operation user has built up a set of dashboards already?  What happens to access to those previously created dashboards?

It would be great to have an "opt-in" switch to turn it on, so that admins can prepare by first exporting the dashboards they want.

Even with the opt-in switch it would be great to be able to provide a tool to dump each ops user's individual kibana dashboard so that they could be imported to the common one.

Comment (portante) [inline README.md]: We also should dove tail this work with the unified dashboard work for non-ops users.

**Comment by jcantrill** [comment]
@richm @portante  This is a reworked PR for https://bugzilla.redhat.com/show_bug.cgi?id=1427254.

I pushed a test version of the image to: docker.io/jcantrill/logging-elasticsearch:bz1427254

which can be utilized by configuring the logging-elasticsearch configmap with:
`openshift.kibana.index.mode: ops_shared`

There are a number of refactorings to make this work.  I think we will additionally need the following:

*operations and .all alias dont seem to map properly. I propose to fix that in subsequent PR
* I think we need to modify hwo the roles are generated per what @richm  found.  User roles to all their projects.  I think this PR will be less then useful. I propose fixing that in subsequent PR.

Please test to see if we think it works as we desire.

**Comment by jcantrill** [comment]
@portante Updated to only 2 modes per your suggestion.  @richm the other change this brings in is to move the allowed actions to action groups defined in the config files.  This will allow us to tweak actions without having to rebuild the plugin for the given roles.  We will need a logging PR to set the defaults and pull in this change.

File: README.md [hunk 5]
```diff
@@ -2,13 +2,28 @@
...
 
 * Dynamically update the SearchGuard ACL based on a user's name
-* Transform kibana index requests to support multitenant deployments
+* Transform kibana index requests to support multitenant deployments for 
+  non-operations users when so configured.
+
+*Note:*
+Previous versions of this plugin created a Kibana profile for each user regardless of their role,
+which is still the default mode of operation.  It is now possible to configure the Kibana index mode 
+to allow operations users to share the Kibana index to store dashboards and visualizations.  It is highly
```
Comment (portante) [inline README.md]: Is it worth defining what defines an "operations" user is?

File: src/main/java/io/fabric8/elasticsearch/plugin/KibanaIndexMode.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+     * The setting that determines the kibana index is used by users.  Valid values are one of the following:
+     * 
+     *   * unique (Default) - Each user gets a unique index for kibana visualizations (e.g. .kibana.USER_UUID)
+     *   * ops_shared       - Users who are in an ops role will share an index (e.g. kibana) while non ops users will 
+     *                        have a unique index (e.g. .kibana.USER_UUID)
+     */
+    static final String OPENSHIFT_KIBANA_INDEX_MODE = "openshift.kibana.index.mode";
+    
+    static final String UNIQUE = "unique";
+    static final String SHARED_OPS = "ops_shared";
```
Comment (portante) [inline src/main/java/io/fabric8/elasticsearch/plugin/KibanaIndexMode.java]: Do you care the constant is `SHARED_OPS` instead of `OPS_SHARED`?

File: README.md (hunk 5 above)
Comment (jcantrill) [inline README.md]: Added link to the file where the SAR is used.

File: src/main/java/io/fabric8/elasticsearch/plugin/KibanaIndexMode.java (hunk 6 above)
Comment (jcantrill) [inline src/main/java/io/fabric8/elasticsearch/plugin/KibanaIndexMode.java]: modified value to 'shared_ops'

**Comment by richm** [comment]
lgtm, but I don't quite follow all of the java changes

**Comment by jcantrill** [comment]
[(3) Ability to seed Kibana dashboard for admins[logging][epic-ois-agl-exp]](https://trello.com/c/O6MCrGUx/454-3-ability-to-seed-kibana-dashboard-for-adminsloggingepic-ois-agl-exp)

**Comment by jcantrill** [comment]
[(3) Ability to seed Kibana dashboard for admins[logging][epic-ois-agl-exp]](https://trello.com/c/O6MCrGUx/454-3-ability-to-seed-kibana-dashboard-for-adminsloggingepic-ois-agl-exp)
